### PR TITLE
Proactive shuffle file cleanup

### PR DESCRIPTION
**Pull request by thinkharderdev**
Cleanup intermediate shuffle files when a job completes. 

In the current setup we just have a timed task that will cleanup shuffle files on a scheduler (using a TTL). This isn't great because we have to balance the TTL against disk space pressure and it's not clear how we can do that effectively. 

This will cleanup intermediate shuffle files once the job completes.

### CODE REVIEW

**Comment by thinkharderdev** [comment]
> Is there any benefit to keeping the intermediate shuffle files?

Not really right now. Once the job is done then they serve no purpose. Theoretically if we got really fancy we could use them as pre-computed results for other queries but that would only be in cases where we do multi-stage aggregations (or aggregations and then filters, etc) and would be monstrously complicated to actually implement.

File: ballista/rust/executor/src/flight_service.rs [hunk 1]
```diff
@@ -163,10 +164,27 @@ impl FlightService for BallistaFlightService {
     ) -> Result<Response<Self::DoActionStream>, Status> {
         let action = request.into_inner();
 
-        let _action =
+        let action =
             decode_protobuf(&action.body.to_vec()).map_err(|e| from_ballista_err(&e))?;
 
-        Err(Status::unimplemented("do_action"))
+        match &action {
+            BallistaAction::FetchPartition { path, .. } => {
```
Comment (mpurins-coralogix) [inline ballista/rust/executor/src/flight_service.rs:171]: `BallistaAction::FetchPartition` -> `BallistaAction::DeletePartition` ?

File: ballista/rust/scheduler/src/state/shuffle_reaper.rs [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+        Self { receiver }
+    }
+
+    pub fn start(self) {
+        tokio::task::spawn(async move {
+            let mut clients: HashMap<String, BallistaClient> = Default::default();
+            let mut receiver = self.receiver;
+            while let Some(locations) = receiver.recv().await {
+                for location in locations {
+                    if let Some(client) = clients.get_mut(&location.executor_meta.id) {
```
Comment (mpurins-coralogix) [inline ballista/rust/scheduler/src/state/shuffle_reaper.rs:22]: I wonder if there could be some kind of 'get or insert' pattern utilised for map here. Found out about entry api (https://doc.rust-lang.org/stable/std/collections/hash_map/struct.HashMap.html#method.entry), but best I could come up with was this which has a downside that executor_id gets always cloned.
```rust
                    let executor_id = location.executor_meta.id.clone();
                    if let Ok(client) = match clients.entry(executor_id) {
                        Entry::Occupied(entry) => Ok(entry.into_mut()),
                        Entry::Vacant(entry) => BallistaClient::try_new(
                            &location.executor_meta.host,
                            location.executor_meta.grpc_port,
                        )
                        .await
                        .map(|client| entry.insert(client)),
                    } {
                        client
                            .delete_partition(&location.path)
                            .await
                            .unwrap_or_else(|e| {
                                error!(
                                "Error deleting partition {:?} from executor {:?}: {:?}",
                                location.path, location.executor_meta.id, e
                            )
                            });
                    } else {
                        error!(
                            "Failed to connect to executor {}",
                            location.executor_meta.id
                        );
                    }
```

File: ballista/rust/scheduler/src/state/shuffle_reaper.rs [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+                        match client.delete_partition(&location.path).await {
```
Comment (mpurins-coralogix) [inline ballista/rust/scheduler/src/state/shuffle_reaper.rs:23]: I think such pattern is more readable when written using `unwrap_or_else`.

So something like this
```rust
match something {
  Ok(()) => (),
  Err(e) => something_else
```
becomes
```rust
something.unwrap_or_else(|| something_else)
```

File: ballista/rust/scheduler/src/state/shuffle_reaper.rs (hunk 2 above)
Comment (thinkharderdev) [reply ballista/rust/scheduler/src/state/shuffle_reaper.rs:22]: yeah, this seems more convoluted to me.

**Pull request closed without merging**